### PR TITLE
Fix Pruning callback and add a few features

### PR DESCRIPTION
**Pull request by carmocca**
## What does this PR do?

Many test fixes
Many refactors

### TODO:

- [x] Make lottery a `Callable` + tests
- [x] Support iterative pruning + tests
- [x] Save pruned model on checkpoint
- [x] Check docs look good
- [x] Sparsity history and tracking
- [x] Allow resampling
- [x] Full model statistics, more verbose level
- [x] Support loading non-permanent pruned checkpoint
- [x] ~~Second chance?~~ edit: left for a future PR (maybe?)
- [x] ~~CHANGELOG~~ edit: No need since pruning hasn't been released yet

## Before submitting
- [x] Was this discussed/approved via a GitHub issue?: Discussed offline
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified
 - [x] **Check that target branch and milestone match!**

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/callbacks/pruning.py [hunk 1]
```diff
@@ -295,96 +278,62 @@ def apply_pruning(self, trainer: 'pl.Trainer', pl_module: LightningModule):
...
+        This function is responsible to check provided ``parameters_to_prune` and `parameters`.
         If parameters_to_prune is None, parameters_to_prune will be generated from all parameters of the model.
         """
+        parameters = parameters or ModelPruning.PARAMETER_NAMES
 
-        is_parameters_to_prune_none = parameters_to_prune is None
         current_modules = [
-            m for m in pl_module.modules()
-            if not isinstance(m, (LightningModule, ModuleDict, ModuleList))
+            m for m in pl_module.modules() if not isinstance(m, (LightningModule, ModuleDict, ModuleList))
```
Comment (tchaton) [inline pytorch_lightning/callbacks/pruning.py:309]: Forgot Sequential there.

File: pytorch_lightning/callbacks/pruning.py [hunk 2]
```diff
@@ -51,18 +46,17 @@
...
         parameters_to_prune: Optional[List[Tuple[nn.Module, str]]] = None,
-        parameter_names: List[str] = ["weight"],
+        parameter_names: Optional[List[str]] = None,
         use_global_unstructured: bool = True,
-        amount: Optional[Union[int, float]] = 0.5,
-        make_pruning_permanent: Optional[bool] = True,
-        use_lottery_ticket_hypothesis: Optional[bool] = True,
+        amount: Union[int, float, Callable] = 0.5,
+        make_pruning_permanent: bool = True,
+        use_lottery_ticket_hypothesis: bool = True,
```
Comment (tchaton) [inline pytorch_lightning/callbacks/pruning.py:59]: Can this become a callback too ?

File: pytorch_lightning/callbacks/pruning.py (hunk 1 above)
Comment (tchaton) [reply pytorch_lightning/callbacks/pruning.py:309]: Let's move this to _MODULE_CONTAINERS = (LightningModule, ModuleDict, ModuleList, Sequential)

File: pytorch_lightning/callbacks/pruning.py [hunk 3]
```diff
@@ -80,77 +79,84 @@ def __init__(
...
         self._use_lottery_ticket_hypothesis = use_lottery_ticket_hypothesis
+        self._resample_parameters = resample_parameters
         self._parameter_names = parameter_names or self.PARAMETER_NAMES
         self._global_kwargs = {}
-        self._initial_parameters_to_prune = None
+        self._original_layers = None
 
-        for param_name in self._parameter_names:
-            if param_name not in self.PARAMETER_NAMES:
+        for name in self._parameter_names:
```
Comment (tchaton) [inline pytorch_lightning/callbacks/pruning.py:145]: IMO, param_name was more explicit.

File: pytorch_lightning/callbacks/pruning.py [hunk 4]
```diff
@@ -215,175 +217,183 @@ def _create_pruning_fn(self, pruning_fn: str, *args, **kwargs):
...
-            pytorch_prune.remove(module, param_name)
-
-    def _resolve_amount(self, current_epoch: int) -> float:
-        if isinstance(self.amount, Callable):
-            amount_fn = self.amount
-            amount = amount_fn(current_epoch)
-        else:
-            amount = self.amount
-        return amount
+            try:
```
Comment (tchaton) [inline pytorch_lightning/callbacks/pruning.py:229]: Not sure about try expect. We just to check for weight attribute.

File: pytorch_lightning/callbacks/pruning.py [hunk 5]
```diff
@@ -215,175 +217,183 @@ def _create_pruning_fn(self, pruning_fn: str, *args, **kwargs):
...
+            2. Train the network for j iterations, arriving at parameters θ_j .
+            3. Prune p% of the parameters in θ_j, creating a mask m.
+            4. Reset the remaining parameters to their values in θ_0, creating the winning ticket f(x; m⊙θ_0).
+
+        This function implements the step 4.
+
+        The ``resample_parameters`` argument can be used to reset the parameters with a new θ_j ∼ D_θ
+        """
+
+        def copy_param(new, old, name: str) -> None:
```
Comment (tchaton) [inline pytorch_lightning/callbacks/pruning.py:256]: Make private.

File: pytorch_lightning/callbacks/pruning.py [hunk 6]
```diff
@@ -215,175 +217,183 @@ def _create_pruning_fn(self, pruning_fn: str, *args, **kwargs):
...
         )
 
-    def apply_pruning(self, trainer: 'pl.Trainer', pl_module: LightningModule):
-        amount = self._resolve_amount(trainer.current_epoch)
-
-        # the user could control the pruning frequency with amount_fn
-        if amount == 0 or amount is None:
-            return
+    @staticmethod
+    def _get_pruned_stats(module: nn.Module, name: str) -> Tuple[int, int]:
```
Comment (tchaton) [inline pytorch_lightning/callbacks/pruning.py:288]: Let's add a property sparsity computed on the entire model parameters.

File: pytorch_lightning/callbacks/pruning.py [hunk 7]
```diff
@@ -215,175 +217,183 @@ def _create_pruning_fn(self, pruning_fn: str, *args, **kwargs):
...
+            stats = [self._get_pruned_stats(m, n) for m, n in self._parameters_to_prune]
 
         if self._use_global_unstructured:
             self._apply_global_pruning(amount)
         else:
             self._apply_local_pruning(amount)
 
-        if self._use_lottery_ticket_hypothesis:
-            self.apply_lottery_ticket_hypothesis()
+        if self._verbose:
```
Comment (tchaton) [inline pytorch_lightning/callbacks/pruning.py:305]: I think we could have 2 levels of verbose. Overall, and detailed.

File: pytorch_lightning/callbacks/pruning.py [hunk 8]
```diff
@@ -215,175 +217,183 @@ def _create_pruning_fn(self, pruning_fn: str, *args, **kwargs):
...
-
-        if self.make_pruning_permanent:
-            self._make_pruning_permanent()
+            # group modules by id. Each entry has a copy of the initial data
+            # and a list of the associated parameter names to prune
+            self._original_layers = {}
+            for i, (module, name) in enumerate(self._parameters_to_prune):
+                id_ = id(module)
+                self._original_layers.setdefault(id_, {"data": deepcopy(module), "names": []})
+                self._original_layers[id_]["names"].append((i, name))
```
Comment (tchaton) [inline pytorch_lightning/callbacks/pruning.py:329]: Why not use self._original_layers[module] = name directly ?

**Review by tchaton — COMMENTED** [review]
Great progress :)

File: pytorch_lightning/callbacks/pruning.py [hunk 9]
```diff
@@ -47,31 +44,33 @@
...
+        pruning_fn: Union[Callable, str],
+        parameters_to_prune: Optional[_PARAM_LIST] = None,
+        parameter_names: Optional[List[str]] = None,
         use_global_unstructured: bool = True,
-        amount: Optional[Union[int, float]] = 0.5,
-        make_pruning_permanent: Optional[bool] = True,
-        use_lottery_ticket_hypothesis: Optional[bool] = True,
+        amount: Union[int, float, Callable[[int], Union[int, float]]] = 0.5,
+        apply_pruning: Union[bool, Callable[[int], bool]] = True,
+        make_pruning_permanent: bool = True,
```
[the remark this answers is not in the document]
Comment (carmocca) [reply pytorch_lightning/callbacks/pruning.py:62]: It is, see `on_save_checkpoint`

File: pytorch_lightning/callbacks/pruning.py (hunk 3 above)
Comment (carmocca) [reply pytorch_lightning/callbacks/pruning.py:145]: Notice this change is internal, users still pass `param_names`. Using `name` internally follow PyTorch's naming

File: pytorch_lightning/callbacks/pruning.py (hunk 4 above)
Comment (carmocca) [reply pytorch_lightning/callbacks/pruning.py:229]: Unfortunately, PyTorch's `remove` does not have a `strict=False` argument and will raise a `ValueError` when not found. So it's the easiest way to check it.

https://github.com/pytorch/pytorch/blob/56034636b9f37b54dacfab8794e93ba6b10bdfc7/torch/nn/utils/prune.py#L1182-L1185

File: pytorch_lightning/callbacks/pruning.py (hunk 5 above)
Comment (carmocca) [reply pytorch_lightning/callbacks/pruning.py:256]: This is a function inside another function. It is private already, can't be accessed from outside.

File: pytorch_lightning/callbacks/pruning.py (hunk 8 above)
Comment (carmocca) [reply pytorch_lightning/callbacks/pruning.py:329]: Because we wan't to remove duplicate modules.

For example:

```python
parameters_to_prune = [
    (a, "weight")
    (b, "weight")
    (a, "bias")
]
```

produces

```python
original_layers{
    id(a): {"data": deepcopy(a), "names": [(0, "weight"), (2, "bias)]},
    id(b): {"data": deepcopy(b), "names": [(1, "weight")]},
}
```

This avoids making duplicate copies of the modules while still having a reference to the original `parameters_to_prune` position. This reference is necessary because `_parameters_to_prune` is what's updated with each pruning

File: pytorch_lightning/callbacks/pruning.py (hunk 7 above)
Comment (carmocca) [reply pytorch_lightning/callbacks/pruning.py:305]: What's the difference between the two?

**Comment by Borda** [comment]
it would be much easier to have each of the 5 features as a separate PR

**Comment by carmocca** [comment]
> it would be much easier to have each of the 5 features as a separate PR

Consider this as a Pruning make-over more than adding individual features. The TODO above was just for me to track pending things

**Comment by Borda** [comment]
> > it would be much easier to have each of the 5 features as a separate PR
> 
> Consider this as a Pruning make-over more than adding individual features. The TODO above was just for me to track pending things

yes but it now because there is some code already, so remove all in one PR and then do this, now the reading with seeing past state is hard...

**Review by justusschock — APPROVED** [review]

**Review by SeanNaren — APPROVED** [review]

**Review by tchaton — APPROVED** [review]
Hey @carmocca, incredible work there !

**Comment by carmocca** [comment]
> [Line 394:13](https://github.com/PyTorchLightning/pytorch-lightning/blob/79400365cffd79c3d57eb330f34bf529691c174d/pytorch_lightning/callbacks/pruning.py#L394): [W503](https://duckduckgo.com/?q=pep8%20W503) line break before binary operator
> [Line 395:13](https://github.com/PyTorchLightning/pytorch-lightning/blob/79400365cffd79c3d57eb330f34bf529691c174d/pytorch_lightning/callbacks/pruning.py#L395): [W503](https://duckduckgo.com/?q=pep8%20W503) line break before binary operator

This is a false positive. This rule should be ignored:

https://github.com/PyTorchLightning/pytorch-lightning/blob/388eeea52d3abe5e8c08197ed06dc9ca9eeefe1c/setup.cfg#L83

File: pyproject.toml [hunk 10]
```diff
@@ -6,7 +6,7 @@ requires = [
 
 [tool.autopep8]
```
Comment (Borda) [inline pyproject.toml:7]: all this can be dropped as we do not use it...

Comment (carmocca) [reply pyproject.toml:7]: ah okay. changed it to see if it had anything to do with pep8speaks. I'll remove it in a separate PR

File: pyproject.toml [hunk 11]
```diff
@@ -6,7 +6,7 @@ requires = [
...
 max_line_length = 120
-ignore = ["W503", "W504", "E402", "E731", "C40", "E741", "F40", "F841"]
+ignore = ["W503", "E402", "E731", "C40", "E741", "F40", "F841"]
```
Comment (carmocca) [inline pyproject.toml:9]:
```suggestion
ignore = ["W503", "W504", "E402", "E731", "C40", "E741", "F40", "F841"]
```
